### PR TITLE
Add JSON schema check when create deployment

### PR DESCRIPTION
**Pull request by SLdragon**

### CODE REVIEW

File: src/iotEdgeExplorer.ts [hunk 1]
```diff
@@ -102,6 +114,23 @@ export class IoTEdgeExplorer extends BaseExplorer {
...
+            const ajv = new Ajv({ allErrors: true });
+            const valid = ajv.validate(schema, JSON.parse(json));
+            if (!valid) {
+                this._outputChannel.show();
+                this.outputLine(Constants.IoTHubEdgeLabel, `There are errors in deployment json file:\n${ajv.errorsText(null, { separator: ",\n" })}`);
+                return false;
+            }
+            return true;
+        } catch (error) {
+            vscode.window.showWarningMessage(`Cannot validate deployment json: ${error.toString()}`);
```
Comment (adashen) [inline src/iotEdgeExplorer.ts]: the message here maybe annoying. User could have correct schema, but met with this warning. Do we really need user aware of it here? Or we just need collect some telemetry data?

File: package.json [hunk 2]
```diff
@@ -700,6 +700,8 @@
   },
   "dependencies": {
     "@azure/event-hubs": "^1.0.8",
+    "ajv": "^6.10.0",
```
Comment (adashen) [inline package.json]: new dependency added. We may need to refresh our third-party notice also.

File: src/iotEdgeExplorer.ts [hunk 3]
```diff
@@ -102,6 +114,23 @@ export class IoTEdgeExplorer extends BaseExplorer {
         }
     }
 
+    private async isValidDeploymentJson(json: string): Promise<boolean> {
+        try {
+            const schema = (await axios.get(Constants.DeploymentJsonSchemaUrl)).data;
+            const ajv = new Ajv({ allErrors: true });
+            const valid = ajv.validate(schema, JSON.parse(json));
```
Comment (adashen) [inline src/iotEdgeExplorer.ts]: JSON.parse(json) may throw error, and we need catch this error to let user know that it is not a json string.

Comment (adashen) [reply src/iotEdgeExplorer.ts]: So the validation includes two parts. 1. verify it is a valid json 2. Verify the schema is correct.

File: src/iotEdgeExplorer.ts [hunk 4]
```diff
@@ -102,6 +114,23 @@ export class IoTEdgeExplorer extends BaseExplorer {
...
     }
 
+    private async isValidDeploymentJson(json: string): Promise<boolean> {
+        try {
+            const schema = (await axios.get(Constants.DeploymentJsonSchemaUrl)).data;
+            const ajv = new Ajv({ allErrors: true });
+            const valid = ajv.validate(schema, JSON.parse(json));
+            if (!valid) {
+                this._outputChannel.show();
+                this.outputLine(Constants.IoTHubEdgeLabel, `There are errors in deployment json file:\n${ajv.errorsText(null, { separator: ",\n" })}`);
```
Comment (adashen) [inline src/iotEdgeExplorer.ts]: when the schema is invalid, we need a more clear way to let user know. So message box here makes sense.

File: src/iotEdgeExplorer.ts [hunk 5]
```diff
@@ -102,6 +114,40 @@ export class IoTEdgeExplorer extends BaseExplorer {
...
+    private async isValidDeploymentJson(jsonStr: string): Promise<boolean> {
+        try {
+            const schema = (await axios.get(Constants.DeploymentJsonSchemaUrl)).data;
+            const ajv = new Ajv({ allErrors: true });
+
+            const json = this.tryParseJson(jsonStr);
+            if (json === null) {
+                TelemetryClient.sendEvent(Constants.IoTHubAIValidateJsonSchemaEvent, { error: "The deployment json file is not a valid json file" });
+                vscode.window.showErrorMessage("The deployment json file is not a valid json file");
+                return false;
```
Comment (adashen) [inline src/iotEdgeExplorer.ts]: how about just throw error here. and the telemetry of error and error message could be sent in catch block.

Comment (SLdragon) [reply src/iotEdgeExplorer.ts]: When deployment file is not a json string, then we will **not continue** to create deployment, so it return false.

However, when we cannot validate json schema ( for example: user's network is poor, or some internal exception throw by the validator), and we will **continue** to create deployment, so it return true.

I think this two types of  error cannot combined together.

File: src/iotEdgeExplorer.ts (hunk 1 above)
Comment (SLdragon) [reply src/iotEdgeExplorer.ts]: Only send telemetry instead

File: package.json (hunk 2 above)
Comment (SLdragon) [reply package.json]: I will add it when this pull request merged into master

File: src/iotEdgeExplorer.ts (hunk 3 above)
Comment (SLdragon) [reply src/iotEdgeExplorer.ts]: Fixed

File: src/iotEdgeExplorer.ts (hunk 4 above)
Comment (SLdragon) [reply src/iotEdgeExplorer.ts]: Use message box instead

File: src/iotEdgeExplorer.ts (hunk 5 above)
Comment (adashen) [reply src/iotEdgeExplorer.ts]: Yes, better to different types of error and in error handling we could handle different types of errors. The sendEvent telemetry in lots of if/else block not looks clean.

File: src/iotEdgeExplorer.ts [hunk 6]
```diff
@@ -116,28 +117,36 @@ export class IoTEdgeExplorer extends BaseExplorer {
...
-                return false;
+                errorMessage = "The deployment json file is not a valid json file";
             }
 
-            const valid = ajv.validate(schema, json);
-            if (!valid) {
-                TelemetryClient.sendEvent(Constants.IoTHubAIValidateJsonSchemaEvent, { error: `There are errors in deployment json file: ${ajv.errorsText(null, { separator: ", " })}` });
-                vscode.window.showErrorMessage(`There are errors in deployment json file: ${ajv.errorsText(null, { separator: ", " })}`);
-                return false;
+            if (json) {
```
Comment (adashen) [inline src/iotEdgeExplorer.ts]: else here?

Comment (SLdragon) [reply src/iotEdgeExplorer.ts]: fixed

File: src/iotEdgeExplorer.ts [hunk 7]
```diff
@@ -152,8 +166,18 @@ export class IoTEdgeExplorer extends BaseExplorer {
             if (!contentJson.modulesContent && contentJson.moduleContent) {
                 contentJson.modulesContent = contentJson.moduleContent;
                 delete contentJson.moduleContent;
-                content = JSON.stringify(contentJson, null, 2);
             }
+
+            try {
+                const isValid = await this.isValidDeploymentJsonSchema(contentJson);
```
Comment (adashen) [inline src/iotEdgeExplorer.ts]: So now we don't collect telemetry whether user has invalid json schema event?

**Pull request merged**